src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs

```diff
@@ -261,43 +251,88 @@ static DiscoveryCriteria NewDiscoveryCriteriaFromSourceAndSettings(IEnumerable<s
-                    handler.HandleLogMessage(TestMessageLevel.Error, exceptionToString);
-
-                    // Send discovery complete. Similar logic is also used in ProxyDiscoveryManager.DiscoverTests.
-                    // Differences:
-                    // Total tests must be zero here since parallel discovery events handler adds the count
-                    // Keep `lastChunk` as null since we don't want a message back to the IDE (discovery didn't even begin)
-                    // Set `isAborted` as true since we want this instance of discovery manager to be replaced
-                    // TODO: the comment above mentions 0 tests but sends -1. Make sense of this.
-                    var discoveryCompleteEventsArgs = new DiscoveryCompleteEventArgs(-1, true);
-                    handler.HandleDiscoveryComplete(discoveryCompleteEventsArgs, null);
-                },
-                TaskContinuationOptions.OnlyOnFaulted);
+            }).ContinueWith(t => HandleError(eventHandler, t), TaskContinuationOptions.OnlyOnFaulted);
 
         EqtTrace.Verbose("ProxyParallelDiscoveryManager.DiscoverTestsOnConcurrentManager: No sources available for discovery.");
     }
 
+    private void HandleError(ITestDiscoveryEventsHandler2 eventHandler, Task t)
+    {
+        // Just in case, the actual discovery couldn't start for an instance. Ensure that
+        // we call discovery complete since we have already fetched a source. Otherwise
+        // discovery will not terminate
+        EqtTrace.Error("ParallelProxyDiscoveryManager: Failed to trigger discovery. Exception: " + t.Exception);
+
+        var handler = eventHandler;
+        var testMessagePayload = new TestMessagePayload { MessageLevel = TestMessageLevel.Error, Message = t.Exception?.ToString() };
+        handler.HandleRawMessage(_dataSerializer.SerializePayload(MessageType.TestMessage, testMessagePayload));
+        handler.HandleLogMessage(TestMessageLevel.Error, t.Exception?.ToString());
+
+        // Send discovery complete. Similar logic is also used in ProxyDiscoveryManager.DiscoverTests.
+        // Differences:
+        // Total tests must be zero here since parallel discovery events handler adds the count
+        // Keep `lastChunk` as null since we don't want a message back to the IDE (discovery didn't even begin)
+        // Set `isAborted` as true since we want this instance of discovery manager to be replaced
+        // TODO: the comment above mentions 0 tests but sends -1. Make sense of this.
+        var discoveryCompleteEventsArgs = new DiscoveryCompleteEventArgs(-1, true);
+        handler.HandleDiscoveryComplete(discoveryCompleteEventsArgs, null);
+    }
+
+    public void InitializeDiscovery(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler, bool skipDefaultAdapters)
+    {
+        throw new NotImplementedException();
```

Comment:
Comment about why it's not implemented.

Reply:
Done, thanks.

Reply:
@nohwnd I see some discrepancy between how we treat `InitializeDiscovery` and `InitializeTestRun`. For the first one we throw a `NotImplementedException` while the second one is just empty. I decided to go ahead and leave the body of `InitializeDiscovery` empty too instead of throwing an exception. Please let me know if you have any concerns on this.